LeaderboardBackend/Services/Impl/LeaderboardService.cs

```diff
@@ -42,4 +43,34 @@ public async Task<CreateLeaderboardResult> CreateLeaderboard(CreateLeaderboardRe
 
         return lb;
     }
+
+    public async Task<RestoreLeaderboardResult> RestoreLeaderboard(long id)
+    {
+        Leaderboard? lb = await applicationContext.Leaderboards.FindAsync([id]);
+
+        if (lb == null)
+        {
+            return new LeaderboardNotFound();
+        }
+
+        if (lb.DeletedAt == null)
+        {
+            return new LeaderboardNeverDeleted();
+        }
+
+        Leaderboard? maybe = await applicationContext.Leaderboards.SingleOrDefaultAsync(board => board.Slug == lb.Slug && board.DeletedAt == null);
+
+        if (maybe != null)
+        {
+            return new RestoreLeaderboardConflict(maybe);
+        }
```

Comment:
Another way to do this would be to simply try to restore it and return the offender if it fails due to a conflict.

Reply:
Can we just keep this? Unless it really becomes a problem.

Reply:
Oh yeah that's true. I'll update that later. Was really tired when I replied; didn't wanna spend even more time changing code around ha

Reply:
@TheTedder Trying to follow the example of that UserService method rn; you know I could get the conflicting board from the PostgresException? Or would I still need to make a DB call?

Reply:
Also; following that pattern; I'd make a `LeaderboardEntityTypeConfig.SLUG_UNIQUE_INDEX` in the entity file. Would I also need to change the `HasFilter` call in the builder definition, too? I.e. edit `LeaderboardEntityTypeConfig`'s definition to:
```cs
public class LeaderboardEntityTypeConfig : IEntityTypeConfiguration<Leaderboard>
{
    public const string SLUG_UNIQUE_INDEX = "ix_leaderboards_slug";

    public void Configure(EntityTypeBuilder<Leaderboard> builder) =>
        builder.HasIndex(l => l.Slug)
            .IsUnique()
            .HasDatabaseName(SLUG_UNIQUE_INDEX); // replaces current `HasFilter` call
}
```

Reply:
I'll take a closer look later when I get home but you don't need to check for which index specifically caused the issue because there's only one unique index on the table, meaning the table doesn't need to be modified.

Reply:
You could do it like this, avoiding an extra DB call if no conflict happens.
```cs
public async Task<RestoreLeaderboardResult> RestoreLeaderboard(long id)
{
    Leaderboard? lb = await applicationContext.Leaderboards.FindAsync(id);

    if (lb == null)
    {
        return new LeaderboardNotFound();
    }

    if (lb.DeletedAt == null)
    {
        return new LeaderboardNeverDeleted();
    }

    lb.DeletedAt = null;

    try
    {
        await applicationContext.SaveChangesAsync();
    }
    catch (DbUpdateException e)
        when(e.InnerException is PostgresException { SqlState: PostgresErrorCodes.UniqueViolation } pgEx)
    {
        Leaderboard conflict = await applicationContext.Leaderboards.SingleAsync(c => c.Slug == lb.Slug && c.DeletedAt == null);
        return new RestoreLeaderboardConflict(conflict);
    }

    return lb;
}
```

Reply:
I think the `return` should be inside the `try`, right? Or more importantly, what's the behaviour if a different exception is thrown?

Reply:
Whether the return statement is inside the try block or not doesn't really make a difference. If an unrelated exception is thrown, the error goes uncaught and the server returns a 500 status code.

Reply:
It be updated